File: users/views/analysis_tools.py

# Django
from django.contrib.auth.mixins import LoginRequiredMixin
import json
from django.db.models import OuterRef, Sum, F, Subquery, Q, Count
from django.db.models.functions import Coalesce, Ceil, NullIf

# Models
from programs.models import BasePart, ReusedPart, Program
from logs.models import Phase, TimeLog

# Utils Constants
from users.utils import TIME_TOTAL_BY_PROGRAM

# Django REST Framework
from rest_framework.views import APIView
from rest_framework.response import Response
from rest_framework import status

# Serializers
from users.serializers import DataProgramAnalysisTools

# Mixins
from users.mixins import UserExistMixin, AllowAccessUserPageMixin


class UserAnalysisToolsView(LoginRequiredMixin, UserExistMixin, AllowAccessUserPageMixin, APIView):

    data = {
        'actual_size': {},
        'defects_removed': {},
        'defects_injected': {},
        'total_time': {},
        'appraisal_COQ': {},
        'total_defects': {}
    }

    def get(self, request, *args, **kwargs):

        # -----> ACTUAL SIZE <-------
        base = BasePart.objects.values('program__pk').filter(program=OuterRef("pk")).annotate(total=Sum('lines_current_base'))
        added = BasePart.objects.values('program__pk').filter(program=OuterRef("pk")).annotate(total=Sum('lines_current_added'))
        deleted = BasePart.objects.values('program__pk').filter(program=OuterRef("pk")).annotate(total=Sum('lines_current_deleted'))
        reused = ReusedPart.objects.values('program__pk').filter(program=OuterRef("pk")).annotate(total=Sum('current_lines'))

        self.data["actual_size"] = Program.objects.values('pk', 'name').filter(programmer=self.user, finish_date__isnull=False).annotate(total=Coalesce(Subquery(base.values('total')) - Subquery(deleted.values('total')) + Subquery(reused.values('total')) + Subquery(added.values('total')), 0))
        # -----> END ACTUAL SIZE <------

        # -----> Defects Removed <--------
        self.data["defects_removed"] = Program.objects.filter(programmer=self.user, finish_date__isnull=False).annotate(
            planning=Count('name', Q(get_defects__phase_removed__name='Planning')),
            design=Count('name', Q(get_defects__phase_removed__name='Design')),
            design_review=Count('name', Q(get_defects__phase_removed__name='Design Review')),
            codification=Count('name', Q(get_defects__phase_removed__name='Codification')),
            codification_review=Count('name', Q(get_defects__phase_removed__name='Codification Review')),
            compilation=Count('name', Q(get_defects__phase_removed__name='Compilation')),
            unit_test=Count('name', Q(get_defects__phase_removed__name='Unit Test')),
            postmortem=Count('name', Q(get_defects__phase_removed__name='Postmortem'))
        ).values('pk', 'name', 'planning', 'design', 'design_review', 'codification', 'codification_review', 'compilation', 'unit_test', 'postmortem').order_by('created_at')
        # -----> END Defects removed <------


        # -----> Defects Injected <-------
        self.data["defects_injected"] = Program.objects.filter(programmer=self.user, finish_date__isnull=False).annotate(
            planning=Count('name', Q(get_defects__phase_injected__name='Planning')),
            design=Count('name', Q(get_defects__phase_injected__name='Design')),
            design_review=Count('name', Q(get_defects__phase_injected__name='Design Review')),
            codification=Count('name', Q(get_defects__phase_injected__name='Codification')),
            codification_review=Count('name', Q(get_defects__phase_injected__name='Codification Review')),
            compilation=Count('name', Q(get_defects__phase_injected__name='Compilation')),
            unit_test=Count('name', Q(get_defects__phase_injected__name='Unit Test')),
            postmortem=Count('name', Q(get_defects__phase_injected__name='Postmortem'))
        ).values('pk', 'name', 'planning', 'design', 'design_review', 'codification', 'codification_review', 'compilation', 'unit_test', 'postmortem').order_by('created_at')
        # ------> END Defects injected <------

        # -----> Total time <--------
        self.data["total_time"] = Program.objects.filter(programmer=self.user, finish_date__isnull=False).annotate(total=TIME_TOTAL_BY_PROGRAM).values('pk', 'name', 'total').order_by('created_at')
        # -----> END total time <------

        # -----> Failure COQ <--------
        # Hago dos subconsultas que retornan el tiempo dedicado en la fase de compilación y de pruebas
        timelogComp = TimeLog.objects.filter(program__pk=OuterRef("pk"), phase__name='Compilation').annotate(time=Ceil(F('delta_time') / 60.0)).values('time')[:1]
        timelogUnit = TimeLog.objects.filter(program__pk=OuterRef("pk"), phase__name='Unit Test').annotate(time=Ceil(F('delta_time') / 60.0)).values('time')[:1]

        self.data["failure_COQ"] = Program.objects.filter(programmer=self.user, finish_date__isnull=False).annotate(total=Coalesce(100 * ((Coalesce(Subquery(timelogComp), 0) + Coalesce(Subquery(timelogUnit), 0)) / NullIf(TIME_TOTAL_BY_PROGRAM, 0)), 0)).values('pk', 'name', 'total').order_by('created_at')
        # ------> END Failure COQ <------

        # ------> Appraisal Cost Of Quality <---------
        # Hago dos subconsultas que retornan el tiempo dedicado en la fase de revisión de diseño y revisión de código
        timelog_design_review = TimeLog.objects.filter(program__pk=OuterRef("pk"), phase__name='Design Review').annotate(time=Ceil(F('delta_time') / 60.0)).values('time')[:1]
        timelog_code_review = TimeLog.objects.filter(program__pk=OuterRef("pk"), phase__name='Codification Review').annotate(time=Ceil(F('delta_time') / 60.0)).values('time')[:1]

        self.data["appraisal_COQ"] = Program.objects.filter(programmer=self.user, finish_date__isnull=False).annotate(total=Coalesce(100 * ((Coalesce(Subquery(timelog_design_review), 0) + Coalesce(Subquery(timelog_code_review), 0)) / NullIf(TIME_TOTAL_BY_PROGRAM, 0)), 0)).values('pk', 'name', 'total').order_by('created_at')
        # -------> END Total defects <-------

        # ------> Total defects <---------
        self.data["total_defects"] = Program.objects.filter(programmer=self.user, finish_date__isnull=False).annotate(total=Count('get_defects')).values('pk', 'name', 'total')
        # ------> END total defects <-------

        return Response(data=self.data, status=status.HTTP_200_OK)


class TestDataProgram(LoginRequiredMixin, UserExistMixin, AllowAccessUserPageMixin, APIView):

    def get(self, request, *args, **kwargs):
        return Response(data=DataProgramAnalysisTools(data=Program.objects.filter(programmer=self.user)).data, status=status.HTTP_200_OK)

File: static/js/analysis_tools.js


const btnGraphicAnalysis = document.getElementsByClassName('btnGraphicAnalysis')

const descriptionGraphic = document.getElementById('descriptionGraphic')
const titleModalGraphic = document.getElementById('titleModalGraphic')

const ctxGraphic = document.getElementById('canvasGraphicAnalysisTools')

const containerCanvas = document.getElementById('container-canvas')

// Loader graphic
const loaderGraphicAnalysisTools = document.getElementById('loaderGraphicAnalysisTools')

var canvas = null

Chart.defaults.global.defaultFontFamily = "Lato";
Chart.defaults.global.defaultFontSize = 18;


for (let i = 0; i < btnGraphicAnalysis.length; i++) {
    btnGraphicAnalysis[i].addEventListener('click', (e) => {
        let graphic = e.target.value
        if (graphic == undefined) graphic = e.target.parentElement.value

        setDataGraphic(JSON.parse(graphic))
    })
}

function setDataGraphic(graphic) {
    
    // Limpiar las graficas que se ha generado anteriormente
    if (canvas != null){
        canvas.clear()
        canvas.destroy()
    }

    ctxGraphic.classList.add('d-none')
    loaderGraphicAnalysisTools.classList.remove('d-none')

    let apiService = new APIService()
    descriptionGraphic.textContent = ""

    titleModalGraphic.textContent = ((graphic.name).toUpperCase()).replace('-', ' ')

    apiService.request(`/users/${graphic.user}/analysis-tools/graphics/`, {}, 'GET').then(response => {
        response.json().then(data => {

            loaderGraphicAnalysisTools.classList.add('d-none')
            ctxGraphic.classList.remove('d-none')

            let labels = []
            let dataGraphic = []
            let colors = []
            let datasets = []

            console.log(graphic.name)
            console.log(data)

            switch (graphic.name) {
                
                case 'actual-size':
                    descriptionGraphic.textContent = "Esta grafica muestra el tamaño de cada programa en líneas de código hasta el momento"

                    data.actual_size.map(size => {
                        labels.push(`${size.name.slice(0, 20)}...`)
                        dataGraphic.push(size.total)
                    })

                    canvas = CreatorChart.createChartLine(ctxGraphic.getContext('2d'), labels, dataGraphic, "Programs (Actual Size)")
                    break;

                case 'defects-removed':
                    console.log("DEFECTS REMOVED")
                    descriptionGraphic.textContent = "Esta grafica muestra la cantidad de defectos removidos por fase de cada programa"

                    labels = ["Planning", "Design", "Design Review", "Codification", "Codification Review", "Compilation", "Unit Test", "Postmortem"]
                    colors = []
                    datasets = []

                    data.defects_removed.map((defect, counter) => {
                        
                        let backgroundColor, borderColor

                        do {
                            [backgroundColor, borderColor] = generateColorRGBA()
                        } while (colors.findIndex(c => c == backgroundColor) != -1);

                        colors.push(backgroundColor)

                        datasets.push({
                            label: defect.name,
                            data: [defect.planning, defect.design, defect.design_review, defect.codification, defect.codification_review, defect.compilation, defect.unit_test, defect.postmortem],
                            backgroundColor: [backgroundColor],
                            borderColor: borderColor,
                            borderWidth: 1.5
                        })
                    })

                    canvas = CreatorChart.createRadarChart(ctxGraphic.getContext('2d'), labels, datasets, "Defects removed")
                    break;

                case 'defects-injected':

                    descriptionGraphic.textContent = "Esta grafica muestra la cantidad de defectos injectados por fase de cada programa"

                    labels = ["Planning", "Design", "Design Review", "Codification", "Codification Review", "Compilation", "Unit Test", "Postmortem"]
                    colors = []
                    datasets = []
                    
                    data.defects_injected.map(defect => {
                        let backgroundColor, borderColor
                        
                        do {
                            [backgroundColor, borderColor] = generateColorRGBA()
                        } while (colors.findIndex(c => c == backgroundColor) != -1);

                        colors.push(backgroundColor)
                        
                        datasets.push({
                            label: defect.name,
                            data: [defect.planning, defect.design, defect.design_review, defect.codification, defect.codification_review, defect.compilation, defect.unit_test, defect.postmortem],
                            backgroundColor: [backgroundColor],
                            borderColor: borderColor,
                            borderWidth: 1.5
                        })
                    })  

                    canvas = CreatorChart.createRadarChart(ctxGraphic.getContext('2d'), labels, datasets, "Defects Injected")
                    break;

                case 'total-time':
                    descriptionGraphic.textContent = "Esta gráfica muestra el total de tiempo que se ha empleado por cada programa"
                    
                    data.total_time.map((time, counter) => {
                        labels.push(`Program #${counter + 1}`)
                        datasets.push(time.total)
                    })
                    
                    canvas = CreatorChart.createChartBar(ctxGraphic, labels, datasets, "Time total per program (Minutes)")

                    break;
                
                case 'failure-cost-of-quality':
                    descriptionGraphic.textContent = "Esta grafica muesta la falla del costo de calidad el cual se basa en el tiempo gastado en la fase de compilación y de pruebas sobre el total del tiempo. (Si el % del tiempo gastado en estas fases es mayor al 40% se está hay una falla en la calidad)."

                    data.failure_COQ.map((failure, counter) => {
                        labels.push(`Program #${counter + 1}`)
                        datasets.push(failure.total)
                    })

                    canvas = CreatorChart.createChartBar(ctxGraphic, labels, datasets, "Failure Cost of Quality (%)")
                    break;

                case 'appraisal-cost-of-quality':
                    
                    descriptionGraphic.textContent = "Esta grafica muestra que porcentaje estamos dedicando a la prevención, es decir, esta grafica muestra el porcentaje de tiempo gastado en la fase de revisión de diseño y revisión de codigo respecto al tiempo total"

                    data.appraisal_COQ.map((appraisal, counter) => {
                        labels.push(`Program #${counter + 1}`)
                        datasets.push(appraisal.total)
                    })

                    canvas = CreatorChart.createChartBar(ctxGraphic, labels, datasets, "Appraisal Cost of Quality (%)")
                    break;
                
                case 'total-defects':

                descriptionGraphic.textContent = "Esta grafica muestra la cantidad de defectos que se han cometido hasta el momento en cada programa"

                data.total_defects.map((defect, counter) => {
                    labels.push(`Program #${counter + 1}`)
                    datasets.push(defect.total)
                })

                canvas = CreatorChart.createChartLine(ctxGraphic, labels, datasets, "Total defects by program")
                break;
                
                default:
                    break;
            }
        })
    })
}

function generateColorRGBA() {

    let colorRed = Math.random() * (255 - 0) + 0
    let colorGreen = Math.random() * (255 - 0) + 0
    let colorBlue = Math.random() * (255 - 0) + 0

    return [`rgba(${colorRed}, ${colorGreen}, ${colorBlue}, 0.2)`, `rgba(${colorRed}, ${colorGreen}, ${colorBlue}, 1)`]
}
